xwiki-platform-core/xwiki-platform-refactoring/xwiki-platform-refactoring-default/src/main/java/org/xwiki/refactoring/internal/DefaultRefactoringConfiguration.java

```diff
@@ -38,13 +44,65 @@
 @Singleton
 public class DefaultRefactoringConfiguration implements RefactoringConfiguration
 {
+    private static final String IS_RECYCLE_BIN_SKIPPING_ACTIVATED_PROPERTY = "isRecycleBinSkippingActivated";
+
+    private static final String PREFIX = "refactoring.";
+
     @Inject
     @Named("refactoring")
-    private ConfigurationSource configurationSource;
+    private ConfigurationSource currentWikiConfigurationSource;
+
+    @Inject
+    @Named("refactoringmainwiki")
+    private ConfigurationSource mainWikiConfigurationSource;
+
+    @Inject
+    @Named("xwikiproperties")
+    private ConfigurationSource xwikiPropertiesSource;
+
+    @Inject
+    private WikiDescriptorManager wikiDescriptorManager;
 
     @Override
     public boolean isRecycleBinSkippingActivated()
     {
-        return this.configurationSource.getProperty("isRecycleBinSkippingActivated", false);
+        return defaultPropertyAccess(IS_RECYCLE_BIN_SKIPPING_ACTIVATED_PROPERTY, false);
+    }
+
+    /**
+     * Look for the property hierarchically.
+     *
+     * The value of the property is looked for in the following places, from the first to the last.
+     * <ul>
+     *     <li>In the {@code Refactoring.Code.RefactoringConfigurationClass} instance of the
+     *     {@code Refactoring.Code.RefactoringConfiguration} object in the current wiki.</li>
+     *     <li>In the {@code Refactoring.Code.RefactoringConfigurationClass} instance of the
+     *     {@code Refactoring.Code.RefactoringConfiguration} object in the main wiki wiki.</li>
+     *     <li>In the wiki properties (the property is prefixed with {@code refactoring.}</li>
+     * </ul>
+     * The search stops when a value is found.
+     * If no value is found, the default value is used.
+     *
+     * @param property the property name
+     * @param defaultValue the default value, used if no property declaration is found in the hierarchy
+     * @param <T> the type of the property
+     * @return the value of the property
+     */
+    private <T> T defaultPropertyAccess(String property, T defaultValue)
```

Comment:
Sounds like you're reimplementing the job of CompositeConfigurationSource: @vmassol would know better that stuff, but I do think you don't need to do that.

Reply:
It's actually what we do in most places but if you wanted you could create a new ConfigurationSource component for your need that implements CompositeConfiguration as it's done at https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-configuration/xwiki-platform-configuration-default/src/main/java/org/xwiki/configuration/internal/DefaultConfigurationSource.java

That would probably be nicer I guess. I guess you could do that in the `@Named("refactoring")`  implementation.

Reply:
IIUC, I will add 3 configuration sources to the composite:
1. currentWikiConfigurationSource
2. mainWikiConfigurationSource
3. xwikiPropertiesSource

In the case of a call to `getProperty` from the main wiki, `currentWikiConfigurationSource` and `mainWikiConfigurationSource` are pointing to the same `xwiki:Refactoring.Code.RefactoringConfiguration` document. Consequently, if the default value is set in this document, it will be checked twice before trying `xwikiPropertiesSource`.

The impact on the performance is probably minimal and the code would be simplified. WDYT?

Reply:
Also, the configuration key would not be prefixed with `refactoring` when calling `xwikiPropertiesSource`.

Reply:
Indeed, interesting point ;) 

As you wish. As I said we don't introduce new composite CS in general so we don't have a best practice about that.

Reply:
> Also, the configuration key would not be prefixed with refactoring when calling xwikiPropertiesSource.

That's a good point. We're still lacking a generic mechanism for that. But this is true for any code that uses the `default` ConfigurationSource.

Reply:
> Also, the configuration key would not be prefixed with refactoring when calling xwikiPropertiesSource

Ok so that making the usage of CompositeConfigurationSource not possible right now. I guess we'd need an improvment issue for that in the future: would be nice to be able to use CompositeConfigurationSource for that kind of usecase.

Reply:
Anyway that's something internal so we could change its implem later if we want. I'm gonna merge it now so we can have it for 12.9.